# 数据分析

## 数据描述

### 其中训练集分为两个部分：

#### `train_July.csv`
7月每一天的订单数据，各个字段含义如下：

 |  字段名称	 |  字段含义	 |  数据示例 | 
  | --- | --- | 
 |  id	 |  订单id的hash值	 |  583411b46a31bcc5d12d4402c928a146 | 
 |  driver_id	 |  司机id的hash值（未出行成功则为一个特殊司机编号） | 	 3e69e17a6e5a726fe44d71896bee4f32 | 
 |  menber_id	 |  乘客id的hash值	  | 6b4d6e4992191fe96b9f27921520d551 | 
 |  create_date | 	 订单创建日期	 |  2017-07-01 | 
 |  create_hour	 |  订单创建时间（0-23小时） | 	 00 | 
 |  status	  | 订单状态：0是未预约成功，1是预约取消，2是出行成功	 |  2 | 
 |  estimate_money	 |  预估行程金额（元） | 	 140.00 | 
 |  estimate_distance | 	 预估行程距离（米） | 	 20099.00 | 
 |  estimate_term | 	 预估行程时间（分钟）	 |  18.00 | 
 |  start_geo_id | 	 起点区域id的hash值	 |  6d7827e8dcfa09497954a31e6f7e6ee6 | 
 |  end_geo_id	 |  终点区域id的hash值 | 	 85e49ded1fa70a7bfa01ab0212a6e538 | 
 

#### `train_Aug.csv`

8月1日至8月7日部分时间段的订单数据，其中奇数日期内选取了偶数小时的订单，偶数日期内选取了奇数小时的订单（例如： 8月1日8时/10时/12时，8月2日7时/9时/11时）。

各字段含义与7月训练数据`train_July.csv`完全相同。


### 测试集分为public set和private set两部分：

#### `test_id_Aug_agg_public5k.csv`
这部分数据给出了8月1日至8月7日的部分时间段，以及起点区域和终点区域，要求选手预测其订单数量。其中奇数日期对应奇数小时，偶数日期对应偶数小时，同训练数据train_Aug.csv在时间上互补。各个字段含义如下：

| 字段名称 |  字段含义	 |  数据示例 | 
 | --- | --- | 
 |  test_id	 |  该条测试用例的id	 |  1 | 
 |  start_geo_id | 	 起点区域id的hash值 | 	 6d7827e8dcfa09497954a31e6f7e6ee6 | 
 |  end_geo_id	 |  终点区域id的hash值 | 	 85e49ded1fa70a7bfa01ab0212a6e538 | 
 |  create_date	 |  订单创建日期	 |  2017-08-01 | 
 |  create_hour	 |  订单创建时间（0-23小时）	 |  01 | 
 

此部分数据为public set。

选手在10月16日至1月12日内需要根据这部分数据提交预测结果，并可以在排行榜查看临时成绩。
 

#### `test_id_Aug_agg_private5k.csv`

数据含义和test_Aug_agg_public5k.csv完全相同，但是数据内容与之不同。

此部分数据为private set，该文件在2018-1-12开放下载。

选手在最后3天（即1月12日至1月15日）需要根据这部分数据提交预测结果，最终比赛结果以此结果为准。
 

>关于数据集的其他说明：

>1. 不同的区域id，代表互不重叠的地图区域，起点区域和终点区域有可能相同

>2. 所有测试集中会有部分数据的实际订单量为0

### 其他数据：

#### `poi.csv`

某一区域内各类POI(point of interest)的数量。

其中第一列为区域id的hash值（与训练及测试数据中的区域对应），后面分别为各类POI的类型及数量，例如：
```
3d99665144344fc090b5b7450ffe72f5,加油站,4,超市,43,住宅区,151,地铁站,4,公交站,36,咖啡厅,22,中餐厅,597,ATM,59,写字楼,47,酒店,45
```

表示该区域内有加油站4个，超市43个，等等

#### `weather.csv`

训练集与测试集所涉及的时间段内的天气情况（有极少部分缺失）。

各个字段含义如下：

 |  字段名称	 |  字段含义	 |  数据示例 | 
  | --- | --- | 
 |  date	 |  日期（精确至半小时） | 	 2017-7-1 0:30 | 
 |  text	 |  天气现象文字 | 	 晴 | 
 |  code	 |  天气现象代码 | 	 1 | 
 |  temperature | 	 温度 | 	 29 | 
 |  feels_like | 	 体感温度	 |  28 | 
 |  pressure | 	 气压 | 	 998 | 
 |  humidity	 |  相对湿度	 |  62 | 
 |  visibility	 |  能见度	 |  9.3 | 
 |  wind_direction	 |  风向文字	|  南 | 
 |  wind_direction_degree	 |  风向角度，范围0-360，0为正北，90为正东，180为正南，270为正西 | 	 200 | 
 |  wind_speed	 |  风速 | 	 8 | 
 |  wind_scale	 |  风力等级	 |  2 | 

### 提交格式 

选手提交的文件须为UTF-8编码(无BOM)的csv格式文件. 数据只需要2列：public set/private set的test_id，和额外加一列并命名为”count”（不含引号），并将预测结果写入这一列。预测结果可以使用浮点数，比如5.3表示比5多一点但不到6. 提交示例如下：

          test_id,count

          0,1

          1,5.3

          ....

 

选手亦可在数据页面下载sample submission作为格式参考.

### 评价指标：

评价指标为MAE(mean absolute error), 即预测值与真实值的绝对误差，并取平均数：
$$ MAE=\frac{1}{n}\sum_{i=1}^{n}|y_i - \hat{y_i}| $$

## 数据预处理

In [1]:
from config import opt
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
print(opt.train_file_0, opt.train_file_1)
print(opt.other_file_0, opt.other_file_1)

#train_data_july = pd.read_csv(opt.train_file_0) # 文件太大170+M，读相对小的8月的数据70M
train_data_aug = pd.read_csv(opt.train_file_1)

other_data_weather = pd.read_csv(opt.other_file_0)

#other_data_poi = pd.read_csv(opt.other_file_1) # 不能直接读取

./data/train/train_July.csv ./data/train/train_Aug.csv
./data/other/weather.csv ./data/other/poi.csv


In [3]:
header = ["geo_id区域ID","加油站","超市","住宅区","地铁站","公交站","咖啡厅","中餐厅","ATM","写字楼","酒店"]
other_data_poi = []
other_data_poi.append(",".join(header)+"\n")

f = open(opt.other_file_1)
count = 0
for i in f.readlines():
    count += 1
    if count == 1:
        print(i)
    i = i[:-1] # 去掉"\n"
    split_to_list = i.split(",")
    concat_to_item = [split_to_list[a] for a in range(0,22,2)]
    other_data_poi.append(",".join(concat_to_item)+"\n")
print(count, len(other_data_poi))
f.close()

f_save = open(opt.other_data_root + "poi_pd.csv","w")
f_save.writelines(other_data_poi)
f_save.close()

829fd9e04ed5f0e2148096737edb8a73,加油站,0,超市,0,住宅区,1,地铁站,0,公交站,2,咖啡厅,0,中餐厅,0,ATM,0,写字楼,0,酒店,0

1234 1235


In [4]:
other_data_poi = pd.read_csv(opt.other_data_root + "poi_pd.csv")

### observe

#### train_data_aug

In [5]:
train_data_aug.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95657 entries, 0 to 95656
Data columns (total 11 columns):
id                   95657 non-null object
driver_id            95657 non-null object
member_id            95657 non-null object
create_date          95657 non-null object
create_hour          95657 non-null int64
status               95657 non-null int64
estimate_money       95657 non-null float64
estimate_distance    95657 non-null float64
estimate_term        95657 non-null float64
start_geo_id         95657 non-null object
end_geo_id           95657 non-null object
dtypes: float64(3), int64(2), object(6)
memory usage: 8.0+ MB


In [6]:
train_data_aug.head()

id                         driver_id  \
0  4274d909ef093f6acf7a10083ed91bda  55f51e0518470d3906ce07e5990a186f   
1  1a7c6820bb2bdaf5c8302770af844444  db8dcdff6c9164152c1974c34715a4b6   
2  0d2261cee2654bd2a250b0458167eb5b  b16a06a5ea94028944a81ad5bbdbb8ca   
3  7944af8f5d1c5e24aca80ed9cf28cb07  d41d8cd98f00b204e9800998ecf8427e   
4  f85e6f1c36ee1887824982c0d49a027a  744a981e300806bc53b0346a98c8f28d   

                          member_id create_date  create_hour  status  \
0  82cab3e2cdaf398987476119a9ebb564  2017-08-01           12       2   
1  6279e4919266998bb5f73c283a5090d3  2017-08-01           12       2   
2  db475cfbf3c4f6dc15b327b0d18713c7  2017-08-01           14       2   
3  737643b4d285bada9f4cbb0e78c16fc6  2017-08-01           14       0   
4  ef8d60c395a297058948794ce93bd4ed  2017-08-01           16       2   

   estimate_money  estimate_distance  estimate_term  \
0           42.00             7324.0           14.0   
1           37.00             5901.0           11.0   
2           32.00             3137.0            6.0   
3           34.53             3672.0            7.0   
4           31.00             4392.0            8.0   

                       start_geo_id                        end_geo_id  
0  aa532f0935c3f7f5d2f1f337125998a0  940ed2bc6835c66f6261c972d4176d25  
1  0a64b01e91cf367a9b016d4f6d42290f  1239d1bfdc532392645d7a37f7469adf  
2  c538ad66d710f99ad0ce951152da36a4  c538ad66d710f99ad0ce951152da36a4  
3  ba20ac2de22cfe373f316eaa256b85fc  ba20ac2de22cfe373f316eaa256b85fc  
4  83fdd3fc77b27bdbfbc1b3a8d7912fbf  3d53327ef089e222f9d4aa6823f60c70

In [7]:
train_data_aug.describe()

create_hour        status  estimate_money  estimate_distance  \
count  95657.000000  95657.000000    95657.000000       95657.000000   
mean      14.862864      1.344679       74.679284       11143.421203   
std        5.784770      0.850609       70.082042       10416.067482   
min        0.000000      0.000000       15.000000           0.000000   
25%       10.000000      1.000000       40.000000        4746.000000   
50%       16.000000      2.000000       58.600000        8536.000000   
75%       20.000000      2.000000       90.850000       14740.000000   
max       23.000000      2.000000     4857.590000      496576.000000   

       estimate_term  
count   95657.000000  
mean       19.397072  
std        14.279750  
min         0.000000  
25%         9.000000  
50%        16.000000  
75%        27.000000  
max       351.000000

In [8]:
status_0 = train_data_aug[train_data_aug["status"]==0]["status"]
status_1 = train_data_aug[train_data_aug["status"]==1]["status"]
status_2 = train_data_aug[train_data_aug["status"]==2]["status"]

estimate_money_0 = train_data_aug[train_data_aug["status"]==0]["estimate_money"]
estimate_money_1 = train_data_aug[train_data_aug["status"]==1]["estimate_money"]
estimate_money_2 = train_data_aug[train_data_aug["status"]==2]["estimate_money"]

estimate_distance_0 = train_data_aug[train_data_aug["status"]==0]["estimate_distance"]
estimate_distance_1 = train_data_aug[train_data_aug["status"]==1]["estimate_distance"]
estimate_distance_2 = train_data_aug[train_data_aug["status"]==2]["estimate_distance"]

estimate_term_0 = train_data_aug[train_data_aug["status"]==0]["estimate_term"]
estimate_term_1 = train_data_aug[train_data_aug["status"]==1]["estimate_term"]
estimate_term_2 = train_data_aug[train_data_aug["status"]==2]["estimate_term"]

In [9]:
print(estimate_money_0.describe(),"\n",
      estimate_money_1.describe(),"\n",
      estimate_money_2.describe())

count    23802.000000
mean        89.906597
std         86.027041
min         15.000000
25%         47.920000
50%         70.630000
75%        106.760000
max       4857.590000
Name: estimate_money, dtype: float64 
 count    15082.000000
mean        76.968533
std         99.434387
min         15.000000
25%         39.000000
50%         57.480000
75%         90.355000
max       3531.160000
Name: estimate_money, dtype: float64 
 count    56773.000000
mean        67.687105
std         48.978198
min         15.000000
25%         37.000000
50%         54.000000
75%         83.000000
max       2796.000000
Name: estimate_money, dtype: float64


预估行程金额

金额越低，越容易预约成功
金额越高，越难预约成功


In [10]:
print(estimate_distance_0.describe(),"\n",
      estimate_distance_1.describe(),"\n",
      estimate_distance_2.describe())

count     23802.000000
mean      11783.147467
std       11197.567954
min           1.000000
25%        5052.000000
50%        8937.500000
75%       15268.000000
max      352578.000000
Name: estimate_distance, dtype: float64 
 count     15082.000000
mean      11212.464660
std       14167.910897
min           0.000000
25%        4378.000000
50%        8089.000000
75%       14374.000000
max      496576.000000
Name: estimate_distance, dtype: float64 
 count     56773.000000
mean      10856.875170
std        8755.594561
min           0.000000
25%        4729.000000
50%        8459.000000
75%       14622.000000
max      372394.000000
Name: estimate_distance, dtype: float64


预约是否成功和距离没有明显关系

In [11]:
print(estimate_term_0.describe(),"\n",
      estimate_term_1.describe(),"\n",
      estimate_term_2.describe())

count    23802.000000
mean        20.424326
std         15.062234
min          0.000000
25%         10.000000
50%         17.000000
75%         28.000000
max        252.000000
Name: estimate_term, dtype: float64 
 count    15082.000000
mean        18.999940
std         16.152469
min          0.000000
25%          8.000000
50%         15.000000
75%         26.000000
max        351.000000
Name: estimate_term, dtype: float64 
 count    56773.000000
mean        19.071897
std         13.362771
min          0.000000
25%          9.000000
50%         16.000000
75%         26.000000
max        267.000000
Name: estimate_term, dtype: float64


预约是否成功和预估时间没有明显关系

#### other_data_weather

In [12]:
other_data_weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1776 entries, 0 to 1775
Data columns (total 12 columns):
date                     1776 non-null object
text                     1776 non-null object
code                     1776 non-null int64
temperature              1776 non-null int64
feels_like               1775 non-null float64
pressure                 1776 non-null int64
humidity                 1776 non-null int64
visibility               1775 non-null float64
wind_direction           1776 non-null object
wind_direction_degree    1776 non-null int64
wind_speed               1776 non-null float64
wind_scale               1776 non-null int64
dtypes: float64(3), int64(6), object(3)
memory usage: 166.6+ KB


In [13]:
other_data_weather.head()

date text  code  temperature  feels_like  pressure  humidity  \
0  2017-7-1 0:30    晴     1           29        28.0       998        62   
1  2017-7-1 1:00    晴     1           28        28.0       998        63   
2  2017-7-1 1:30    晴     1           28        28.0       998        65   
3  2017-7-1 2:00    晴     1           28        27.0       998        66   
4  2017-7-1 2:30    晴     1           28        27.0       998        67   

   visibility wind_direction  wind_direction_degree  wind_speed  wind_scale  
0         9.3              南                    200        9.00           2  
1         7.7             西南                    233        7.92           2  
2         7.7              南                    197        7.20           2  
3         7.7             西南                    233        9.00           2  
4         6.6              南                    180        7.20           2

In [14]:
other_data_weather.describe()

code  temperature   feels_like     pressure     humidity  \
count  1776.000000  1776.000000  1775.000000  1776.000000  1776.000000   
mean      4.052928    28.350225    27.852958   999.824324    68.809122   
std       2.937952     3.637067     3.597618     4.185097    14.379725   
min       0.000000    20.000000    19.000000   990.000000    25.000000   
25%       4.000000    26.000000    26.000000   997.000000    60.000000   
50%       4.000000    28.000000    28.000000  1000.000000    69.000000   
75%       4.000000    31.000000    31.000000  1003.000000    80.000000   
max      15.000000    38.000000    37.000000  1010.000000    96.000000   

        visibility  wind_direction_degree   wind_speed   wind_scale  
count  1775.000000            1776.000000  1776.000000  1776.000000  
mean     10.682986           10279.032095     8.551689     1.913851  
std       8.364376          100076.127726     4.678441     0.778521  
min       0.700000               0.000000     0.000000     0.000000  
25%       4.500000              76.000000     5.400000     1.000000  
50%       7.900000             169.000000     7.560000     2.000000  
75%      14.750000             222.000000    10.800000     2.000000  
max      35.000000          999017.000000    39.240000     6.000000

In [15]:
err = other_data_weather[other_data_weather["wind_direction_degree"] > 360]
err

date text  code  temperature  feels_like  pressure  humidity  \
180   2017-7-4 20:30    阴     9           27        27.0      1003        78   
183   2017-7-4 22:00    阴     9           26        26.0      1003        82   
430   2017-7-10 5:00   多云     4           23        22.0       993        79   
471   2017-7-11 2:00    晴     1           26        25.0       996        69   
472   2017-7-11 2:30    晴     1           25        25.0       996        72   
473   2017-7-11 3:00    晴     1           24        24.0       996        77   
891   2017-7-20 5:30   多云     4           27        27.0       997        87   
1076  2017-7-24 2:30   多云     4           26        25.0       999        85   
1176  2017-7-26 5:00    阴     9           25        24.0      1003        64   
1177  2017-7-26 5:30   多云     4           24        24.0      1003        66   
1178  2017-7-26 6:00   多云     4           24        24.0      1003        66   
1221  2017-7-27 4:00    阴     9           21        21.0      1002        84   
1260  2017-7-28 0:00   多云     4           25        25.0      1006        81   
1262  2017-7-28 1:00   多云     4           25        24.0      1007        85   
1263  2017-7-28 1:30   多云     4           24        24.0      1007        86   
1265  2017-7-28 2:30   多云     4           24        23.0      1007        89   
1682   2017-8-6 1:00   多云     4           26        25.0       997        65   
1730   2017-8-7 1:00    晴     1           27        26.0       997        64   

      visibility wind_direction  wind_direction_degree  wind_speed  wind_scale  
180          5.1              北                 999017        0.00           0  
183          4.2              北                 999017        0.00           0  
430         23.2              北                 999017        0.00           0  
471         15.7              北                 999017        0.72           0  
472         13.6              北                 999017        0.00           0  
473         13.6              北                 999017        0.00           0  
891          2.6              北                 999017        0.00           0  
1076         2.3              北                 999017        0.00           0  
1176        18.6              北                 999017        0.00           0  
1177        10.8              北                 999017        0.00           0  
1178        10.8              北                 999017        0.72           0  
1221         7.8              北                 999017        0.00           0  
1260         7.3              北                 999017        0.00           0  
1262         6.5              北                 999017        0.00           0  
1263         5.5              北                 999017        0.72           0  
1265         5.3              北                 999017        0.00           0  
1682        35.0              北                 999017        0.00           0  
1730        20.0              北                 999017        0.00           0

风向角度，范围0-360，这里有999017超过360度了，去掉

（可能是读取错位，可以修一下，暂时去掉吧）

In [16]:
other_data_weather = other_data_weather[other_data_weather["wind_direction_degree"] <= 360]

#### other_data_poi

In [17]:
other_data_poi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1234 entries, 0 to 1233
Data columns (total 11 columns):
geo_id区域ID    1234 non-null object
加油站           1234 non-null int64
超市            1234 non-null int64
住宅区           1234 non-null int64
地铁站           1234 non-null int64
公交站           1234 non-null int64
咖啡厅           1234 non-null int64
中餐厅           1234 non-null int64
ATM           1234 non-null int64
写字楼           1234 non-null int64
酒店            1234 non-null int64
dtypes: int64(10), object(1)
memory usage: 106.1+ KB


In [18]:
other_data_poi.head()

geo_id区域ID  加油站  超市  住宅区  地铁站  公交站  咖啡厅  中餐厅  ATM  \
0  829fd9e04ed5f0e2148096737edb8a73    0   0    1    0    2    0    0    0   
1  3d99665144344fc090b5b7450ffe72f5    4  43  151    4   36   22  597   59   
2  f524dc6b89fcb9e4a9688c0cb7b2bdb7    0   0    0    0    0    0    0    0   
3  0613fd690240317177a03a0b26ca8591    4  32   50    3   22    6  509   20   
4  687ea595a01fdc18ad438f0d0a325102    0   3    1    0    6    0    6    0   

   写字楼  酒店  
0    0   0  
1   47  45  
2    0   0  
3    9  27  
4    0   1

In [19]:
other_data_poi.describe()

加油站           超市          住宅区          地铁站          公交站  \
count  1234.000000  1234.000000  1234.000000  1234.000000  1234.000000   
mean      0.882496     6.016207    16.302269     0.411669    10.903566   
std       1.677571    11.771606    43.511875     1.179327    16.240526   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     1.000000     1.000000     0.000000     6.000000   
75%       1.000000     6.000000     8.750000     0.000000    14.000000   
max      23.000000   172.000000   324.000000    10.000000   202.000000   

               咖啡厅          中餐厅          ATM          写字楼           酒店  
count  1234.000000  1234.000000  1234.000000  1234.000000  1234.000000  
mean      2.605348    77.406807     5.939222     5.363047     7.688006  
std      14.551271   181.185386    17.572799    21.167952    23.310397  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     4.000000     0.000000     0.000000     0.000000  
75%       0.000000    34.750000     2.000000     1.000000     4.000000  
max     334.000000   900.000000   189.000000   310.000000   279.000000

不标准，修成前面有header的形式

In [20]:
other_data_poi.columns

Index(['geo_id区域ID', '加油站', '超市', '住宅区', '地铁站', '公交站', '咖啡厅', '中餐厅', 'ATM',
       '写字楼', '酒店'],
      dtype='object')

### plot

In [21]:
from pyecharts import Scatter

scatter = Scatter("预估距离和预估时间的关系")
scatter.add("0未预约成功", estimate_distance_0[:100], estimate_term_0[:100])
scatter.add("1预约取消", estimate_distance_1[:100], estimate_term_1[:100])
scatter.add("2出行成功", estimate_distance_2[:100], estimate_term_2[:100])
scatter

取消预约的，更离散

In [22]:
scatter = Scatter("预估金额和预估时间的关系")
scatter.add("0未预约成功", estimate_money_0[:100], estimate_term_0[:100])
scatter.add("1预约取消", estimate_money_1[:100], estimate_term_1[:100])
scatter.add("2出行成功", estimate_money_2[:100], estimate_term_2[:100])
scatter

In [25]:
scatter = Scatter("预估金额和预估时间的关系")
scatter.add("0未预约成功", estimate_money_0[:100], estimate_distance_0[:100])
scatter.add("1预约取消", estimate_money_1[:100], estimate_distance_1[:100])
scatter.add("2出行成功", estimate_money_2[:100], estimate_distance_2[:100])
scatter

预约失败，同一金额，时间长的多